### ISSUE TITLE
Valid matches are reported as unreachable

### ISSUE THREAD

**Issue by CaselIT** [microsoft/pyright#11088]
**Describe the bug**
A clear and concise description of the behavior you are seeing and the expected behavior along with steps to reproduce it.

Matching a class instance using string cases (or I guess other literals) causes pyright to report that a case block is unreachable, even if at runtime it works correctly, the object overrides `__eq__`.
Also from my understanding the specification allows such a case, by stating

>In simple terms, LITERAL will succeed only if <subject> == LITERAL. For the singletons None, True and False, the [is](https://docs.python.org/3/reference/expressions.html#is) operator is used.

https://docs.python.org/3/reference/compound_stmts.html#literal-patterns

**Code or Screenshots**

```python
class X:
    __match_args__ = ("a",)

    def __init__(self, x, y, a):
        self.a = a

    def __eq__(self, value: object) -> bool:
        return self.a == value.a if isinstance(value, X) else self.a == value

def go_match_str(a: X) -> int | str:
    match a:
        case "a":
            return 1
        case "b":
            return 2
        case _:
            return "nope"

def go_match_X(a: X) -> int | str:
    match a:
        case X("a"):
            return 1
        case X("b"):
            return 2
        case _:
            return "nope"

def go_if(a: X) -> int | str:
    if a == "a":
        return 1
    elif a == "b":
        return 2
    else:
        return "nope"


xs = [X(1, 1, "a"), X(2, 2, "b"), X(0, 0, "c")]
assert [go_match_str(x) for x in xs] == [1, 2, "nope"]
assert [go_match_X(x) for x in xs] == [1, 2, "nope"]
assert [go_if(x) for x in xs] == [1, 2, "nope"]
```

the code above triggers no assertion errors but accorting to pyright go_match_str will only match the wildcard pattern:

<img width="394" height="162" alt="Image" src="https://github.com/user-attachments/assets/c291f35a-b0c5-4df4-a613-9882f5098784" />

**VS Code extension or command-line**
Are you running pyright as a VS Code extension, a language server in another editor, integrated into Pylance, or the command-line tool? Which version?

the above image is via pylance, but enabling `reportUnreachable = true` pytight on the above code also reports errors:
>13:13 - error: Type analysis indicates code is unreachable
>15:13 - error: Type analysis indicates code is unreachable

version is pyright 1.1.407
